### ISSUE TITLE
Improve validation schema for list_or_dict fields

### ISSUE THREAD

**Issue by dnephin** [docker/compose#2008]
Right now `extra_hosts` and `labels` use "one of: object or array", but no validation is done on the contents of the `object`.

I think this can be improved by creating a schema for the object that uses `patternProperties`,  so that it will fail if someone uses nested objects, or arrays within the object.

```json
{
  "type": "object",
  "patternProperties": {"type": "string"},
}
```

**Comment by dnephin** [comment]
Apparently you've already have a fix for  this in #2000!

**Comment by dnephin** [comment]
I've updated this issue to be about the `list_or_dict` schema, since it has the same issue, and isn't fixed by #2000
